dockstore-webservice/src/main/java/io/dockstore/webservice/resources/WorkflowResource.java

```diff
@@ -906,6 +950,355 @@ private void updateInfo(Workflow oldWorkflow, Workflow newWorkflow) {
+        }
+        return publishedDeposit;
+    }
+
+    /**
+     * Register a Zenodo DOI for the workflow version
+     * @param workflow    workflow for which DOI is registered
+     * @param workflowVersion workflow version for which DOI is registered
+     */
+    private void registerZenodoDOIForWorkflow(String zenodoAccessToken, Workflow workflow, WorkflowVersion workflowVersion) {
+        ApiClient zendoClient = new ApiClient();
+        // for testing, either 'https://sandbox.zenodo.org/api' or 'https://zenodo.org/api' is the first parameter
+        String zenodoUrlApi = zenodoUrl + "/api";
+        zendoClient.setBasePath(zenodoUrlApi);
+        zendoClient.setApiKey(zenodoAccessToken);
+
+        DepositsApi depositApi = new DepositsApi(zendoClient);
+        ActionsApi actionsApi = new ActionsApi(zendoClient);
+
+        Deposit deposit = new Deposit();
+        Deposit returnDeposit;
+
+        checkForExistingDOIForWorkflowVersion(workflowVersion);
+
+        String latestWorkflowVersionDOIURL = getAnExistingDOIForWorkflow(workflow);
+
+        int depositionID = 0;
+        DepositMetadata depositMetadata = null;
+
+        if (latestWorkflowVersionDOIURL == null || latestWorkflowVersionDOIURL.isEmpty()) {
+            try {
+                // No DOI has been assigned to any version of the workflow yet
+                // So create a new deposit which will enable creation of a new
+                // concept DOI and new version DOI
+                returnDeposit = depositApi.createDeposit(deposit);
+                depositionID = returnDeposit.getId();
+                depositMetadata = returnDeposit.getMetadata();
+
+                fillInMetadata(depositMetadata, workflow, workflowVersion);
+
+            } catch (ApiException e) {
+                LOG.error("Could not create deposition on Zenodo. Error is " + e.getMessage(), e);
+                throw new CustomWebApplicationException("Could not create deposition on Zenodo. "
+                        + "Error is " + e.getMessage(), HttpStatus.SC_BAD_REQUEST);
+            }
+        } else {
+            String depositIdStr = latestWorkflowVersionDOIURL.substring(latestWorkflowVersionDOIURL.lastIndexOf(".") + 1).trim();
+            int depositId = Integer.parseInt(depositIdStr);
+            try {
+                // A DOI was assigned to a workflow version so we will
+                // use the ID associated with the workflow version DOI
+                // to create a new workflow version DOI
+                returnDeposit = actionsApi.newDepositVersion(depositId);
+                // The response body of this action is NOT the new version deposit,
+                // but the original resource. The new version deposition can be
+                // accessed through the "latest_draft" under "links" in the response body.
+                Object links = returnDeposit.getLinks();
+                String depositURL = (String)((LinkedHashMap)links).get("latest_draft");
```

Comment:
Is `latest_draft` guaranteed to exist?

Reply:
The Zenodo documentation https://developers.zenodo.org/#discard doesn't specifically say so, but I assume so since I am creating a new deposition and the documentation says the new version deposition can be accessed through the "latest_draft" under "links" in the response body.